src/test/java/com/google/devtools/build/lib/rules/cpp/CcLibraryConfiguredTargetTest.java

```diff
@@ -1685,4 +1685,139 @@ public void testImplicitOutputsWhitelistOnWhitelist() throws Exception {
+            "    packages = ['//...'],",
+            ")",
+            "filegroup(",
+            "    name = 'srcs',",
+            "    srcs = glob(['**']),",
+            "    visibility = ['//tools/allowlists:__pkg__'],",
+            ")");
+  }
+
+  @Test
+  public void testDynamicLinkTwiceAfterTransition() throws Exception {
+    AnalysisMock.get()
+        .ccSupport()
+        .setupCcToolchainConfig(
+            mockToolsConfig,
+            CcToolchainConfig.builder()
+                .withFeatures(
+                    CppRuleClasses.COPY_DYNAMIC_LIBRARIES_TO_BINARY,
+                    CppRuleClasses.SUPPORTS_DYNAMIC_LINKER));
+
+    prepareCustomTransition();
+
+    scratch.file(
+        "transition/BUILD",
+        "load(':custom_transition.bzl', 'apply_custom_transition')",
+        "cc_binary(",
+        "    name = \"main\",",
+        "    srcs = [\"main.cc\"],",
+        "    linkstatic = 0,",
+        "    deps = [",
+        "        \"dep1\",",
+        "        \"dep2\",",
+        "    ],",
+        ")",
+        "",
+        "apply_custom_transition(",
+        "    name = \"dep1\",",
+        "    deps = [",
+        "        \":dep2\",",
+        "    ],",
+        ")",
+        "",
+        "cc_library(",
+        "    name = \"dep2\",",
+        "    srcs = [\"test.cc\"],",
+        "    hdrs = [\"test.h\"],",
+        ")");
+
+    checkError("//transition:main", "built in a different configuration");
+  }
+
+  @Test
+  public void testDynamicLinkUniqueAfterTransition() throws Exception {
+    AnalysisMock.get()
+            .ccSupport()
+            .setupCcToolchainConfig(
+                    mockToolsConfig,
+                    CcToolchainConfig.builder()
+                            .withFeatures(
+                                    CppRuleClasses.COPY_DYNAMIC_LIBRARIES_TO_BINARY,
+                                    CppRuleClasses.SUPPORTS_DYNAMIC_LINKER));
+
+    prepareCustomTransition();
+
+    scratch.file(
+            "transition/BUILD",
+            "load(':custom_transition.bzl', 'apply_custom_transition')",
+            "cc_binary(",
+            "    name = \"main\",",
+            "    srcs = [\"main.cc\"],",
+            "    linkstatic = 0,",
+            "    deps = [",
+            "        \"dep2\",",
+            "        \"dep3\",",
+            "    ],",
+            ")",
+            "",
+            "apply_custom_transition(",
+            "    name = \"dep1\",",
+            "    deps = [",
+            "        \":dep2\",",
+            "    ],",
+            ")",
+            "",
+            "cc_library(",
+            "    name = \"dep2\",",
+            "    srcs = [\"test.cc\"],",
+            "    hdrs = [\"test.h\"],",
+            ")",
+            "",
+            "cc_library(",
+            "    name = \"dep3\",",
+            "    srcs = [\"other_test.cc\"],",
+            "    hdrs = [\"other_test.h\"],",
+            ")");
+
+
```

Comment:
Here you have to do getConfiguredTarget("//transition:main") then do assertNoEvents(). Otherwise 100% there will be no events because we didn't try to analyze the target.